tools/nodejs_api/src_cpp/node_query_result.cpp

```diff
@@ -0,0 +1,107 @@
+    Napi::HandleScope scope(env);
+
+    Napi::Function t = DefineClass(env, "NodeQueryResult", {
+       InstanceMethod("close", &NodeQueryResult::Close),
+       InstanceMethod("all", &NodeQueryResult::All),
+       InstanceMethod("getNext", &NodeQueryResult::GetNext),
+       InstanceMethod("hasNext", &NodeQueryResult::HasNext),
+    });
+
+    constructor = Napi::Persistent(t);
+    constructor.SuppressDestruct();
+
+    exports.Set("NodeQueryResult", t);
+    return exports;
+}
+
+NodeQueryResult::NodeQueryResult(const Napi::CallbackInfo& info) : Napi::ObjectWrap<NodeQueryResult>(info)  {}
+
+void NodeQueryResult::SetQueryResult(shared_ptr<kuzu::main::QueryResult> & inputQueryResult) {
+    this->queryResult = inputQueryResult;
+}
+
+void NodeQueryResult::Close(const Napi::CallbackInfo& info) {
+    Napi::Env env = info.Env();
+    Napi::HandleScope scope(env);
+
+    try {
+        this->queryResult.reset();
+    } catch(const std::exception &exc) {
+        Napi::TypeError::New(env, "Unsuccessful queryResult close: " + std::string(exc.what())).ThrowAsJavaScriptException();
+    }
+
+}
+
+Napi::Value NodeQueryResult::All(const Napi::CallbackInfo& info) {
+    Napi::Env env = info.Env();
+    Napi::HandleScope scope(env);
+
+    if (info.Length()!=1 || !info[0].IsFunction()) {
+        Napi::TypeError::New(env, "All needs a callback").ThrowAsJavaScriptException();
+        return Napi::Object::New(env);
+    }
+
+    Function callback = info[0].As<Function>();
+    try {
+        AllAsyncWorker* asyncWorker = new AllAsyncWorker(callback, queryResult);
+        asyncWorker->Queue();
+    } catch(const std::exception &exc) {
+        Napi::TypeError::New(env, "Unsuccessful all callback: " + std::string(exc.what())).ThrowAsJavaScriptException();
+    }
+    return info.Env().Undefined();
+}
+
+Napi::Value NodeQueryResult::GetNext(const Napi::CallbackInfo& info) {
+    Napi::Env env = info.Env();
+    Napi::HandleScope scope(env);
+
+    if (info.Length()!=1 || !info[0].IsFunction()) {
+        Napi::TypeError::New(env, "Each needs a callback").ThrowAsJavaScriptException();
+        return Napi::Object::New(env);
+    }
+
+    Function eachCallback = info[0].As<Function>();
+
+    if (!queryResult->hasNext()){
+        Napi::TypeError::New(env, "Cannot call getNext queryResult is empty").ThrowAsJavaScriptException();
+        return Napi::Object::New(env);
+    }
+
+    try {
+        EachSingleAsyncWorker * asyncWorker = new EachSingleAsyncWorker(eachCallback, queryResult, 1);
+        asyncWorker->Queue();
+    } catch(const std::exception &exc) {
+        Napi::TypeError::New(env, "Unsuccessful getNext callback: " + std::string(exc.what())).ThrowAsJavaScriptException();
+    }
+    return info.Env().Undefined();
+}
+
+Napi::Value NodeQueryResult::HasNext(const Napi::CallbackInfo& info) {
+    Napi::Env env = info.Env();
+    Napi::HandleScope scope(env);
+
+    if (info.Length()!=0) {
```

Comment:
Consider remove these checks in C++ and do these checks in JS, so that if the input is not valid, just terminate early instead of propagating to C++.